File: app/src/main/java/me/ykrank/s1next/data/api/model/collection/Posts.java
package me.ykrank.s1next.data.api.model.collection;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.base.Objects;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;

import me.ykrank.s1next.data.api.model.Account;
import me.ykrank.s1next.data.api.model.Post;
import me.ykrank.s1next.data.api.model.Thread;
import me.ykrank.s1next.data.db.BlackListDbWrapper;
import me.ykrank.s1next.data.db.dbmodel.BlackList;
import me.ykrank.s1next.util.StringUtil;

@SuppressWarnings("UnusedDeclaration")
@JsonIgnoreProperties(ignoreUnknown = true)
public final class Posts extends Account {

    @JsonProperty("thread")
    private Thread postListInfo;

    @JsonProperty("threadsortshow")
    private ThreadAttachment threadAttachment;

    @JsonProperty("postlist")
    private List<Post> postList;

    public Thread getPostListInfo() {
        return postListInfo;
    }

    public void setPostListInfo(Thread postListInfo) {
        this.postListInfo = postListInfo;
    }

    public ThreadAttachment getThreadAttachment() {
        return threadAttachment;
    }

    public void setThreadAttachment(ThreadAttachment threadAttachment) {
        this.threadAttachment = threadAttachment;
    }

    public List<Post> getPostList() {
        return postList;
    }

    /**
     * @see #filterPost(Post)
     */
    public static List<Post> filterPostList(final List<Post> oPosts) {
        List<Post> posts = new ArrayList<>();
        for (Post post : oPosts) {
            Post fPost = filterPost(post);
            if (fPost != null) {
                posts.add(fPost);
            }
        }
        return posts;
    }

    /**
     * 对数据源进行处理，标记黑名单用户<br>
     * 如果修改了过滤状态，则会返回不同的对象
     */
    @Nullable
    public static Post filterPost(final Post post) {
        Post nPost = post;
        BlackListDbWrapper blackListWrapper = BlackListDbWrapper.getInstance();
        BlackList blackList = blackListWrapper.getBlackListDefault(Integer.valueOf(post.getAuthorId()), post.getAuthorName());
        if (blackList == null || blackList.post == BlackList.NORMAL) {
            if (post.isHide()) {
                nPost = post.clone();
                nPost.setHide(false);
            }
        } else if (blackList.post == BlackList.DEL_POST) {
            nPost = null;
        } else if (blackList.post == BlackList.HIDE_POST) {
            if (!post.isHide()) {
                nPost = post.clone();
                nPost.setHide(true);
            }
            nPost.setRemark(blackList.remark);
        }
        return nPost;
    }

    public void setPostList(List<Post> postList) {
        this.postList = filterPostList(postList);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        Posts posts = (Posts) o;
        return Objects.equal(postListInfo, posts.postListInfo) &&
                Objects.equal(threadAttachment, posts.threadAttachment) &&
                Objects.equal(postList, posts.postList);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(super.hashCode(), postListInfo, threadAttachment, postList);
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class ThreadAttachment {

        @JsonProperty("threadsortname")
        private String title;

        @JsonProperty("optionlist")
        private ArrayList<Info> infoList;

        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public ArrayList<Info> getInfoList() {
            return infoList;
        }

        public void setInfoList(ArrayList<Info> infoList) {
            this.infoList = infoList;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ThreadAttachment that = (ThreadAttachment) o;
            return Objects.equal(title, that.title) &&
                    Objects.equal(infoList, that.infoList);
        }

        @Override
        public int hashCode() {
            return Objects.hashCode(title, infoList);
        }

        @JsonIgnoreProperties(ignoreUnknown = true)
        public static final class Info implements Parcelable {

            public static final Parcelable.Creator<Info> CREATOR = new Parcelable.Creator<Info>() {

                @Override
                public Info createFromParcel(Parcel source) {
                    return new Info(source);
                }

                @Override
                public Info[] newArray(int size) {
                    return new Info[size];
                }
            };

            @JsonIgnore
            private final String label;

            @JsonIgnore
            private final String value;

            @JsonCreator
            @SuppressWarnings("UnusedDeclaration")
            public Info(@JsonProperty("title") String label,
                        @JsonProperty("value") String value,
                        @JsonProperty("unit") String unit) {
                this.label = label;
                this.value = StringUtil.unescapeNonBreakingSpace(value)
                        + StringUtils.defaultString(unit);
            }

            private Info(Parcel source) {
                label = source.readString();
                value = source.readString();
            }

            public String getLabel() {
                return label;
            }

            public String getValue() {
                return value;
            }

            @Override
            public boolean equals(Object o) {
                if (this == o) return true;
                if (o == null || getClass() != o.getClass()) return false;
                Info info = (Info) o;
                return Objects.equal(label, info.label) &&
                        Objects.equal(value, info.value);
            }

            @Override
            public int hashCode() {
                return Objects.hashCode(label, value);
            }

            @Override
            public int describeContents() {
                return 0;
            }

            @Override
            public void writeToParcel(Parcel dest, int flags) {
                dest.writeString(label);
                dest.writeString(value);
            }
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/data/api/model/collection/Posts.kt
package me.ykrank.s1next.data.api.model.collection

import androidx.annotation.WorkerThread
import com.fasterxml.jackson.annotation.*
import com.github.ykrank.androidtools.util.StringUtil
import me.ykrank.s1next.data.api.model.Account
import me.ykrank.s1next.data.api.model.Post
import me.ykrank.s1next.data.api.model.Thread
import me.ykrank.s1next.data.api.model.Vote
import me.ykrank.s1next.data.db.BlackListDbWrapper
import me.ykrank.s1next.data.db.BlackWordDbWrapper
import me.ykrank.s1next.data.db.dbmodel.BlackList
import me.ykrank.s1next.data.db.dbmodel.BlackWord
import org.apache.commons.lang3.StringUtils
import paperparcel.PaperParcel
import paperparcel.PaperParcelable
import java.util.*

@JsonIgnoreProperties(ignoreUnknown = true)
class Posts : Account {

    @JsonIgnore
    var postListInfo: Thread? = null
        @JsonSetter("thread") set(p) {
            this.postList.forEach {
                if (p?.author == it.authorName) {
                    it.isOpPost = true
                }
            }
            field = p
        }

    @JsonProperty("threadsortshow")
    var threadAttachment: ThreadAttachment? = null

    @JsonIgnore
    var postList: List<Post> = listOf()

    @JsonProperty("special_poll")
    val vote: Vote? = null

    constructor() {}

    @JsonCreator
    constructor(@JsonProperty("special_trade") trade: Map<Int, Any>?,
                @JsonProperty("postlist") postList: List<Post>?) {
        this.postList = filterPostList(postList)
        if (trade != null && postList != null && postList.isNotEmpty()) {
            val post = postList[0]
            if (trade.containsKey(post.id + 1)) {
                post.isTrade = true
            }
        }
    }

    fun initCommentCount(commentCountMap: Map<Int, Int?>) {
        this.postList.forEach {
            if (commentCountMap.containsKey(it.id)) {
                it.rates = listOf()
            }
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        if (!super.equals(other)) return false

        other as Posts

        if (postListInfo != other.postListInfo) return false
        if (threadAttachment != other.threadAttachment) return false
        if (postList != other.postList) return false
        if (vote != other.vote) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + (postListInfo?.hashCode() ?: 0)
        result = 31 * result + (threadAttachment?.hashCode() ?: 0)
        result = 31 * result + postList.hashCode()
        result = 31 * result + (vote?.hashCode() ?: 0)
        return result
    }


    @JsonIgnoreProperties(ignoreUnknown = true)
    class ThreadAttachment {

        @JsonProperty("threadsortname")
        var title: String? = null

        @JsonProperty("optionlist")
        var infoList: ArrayList<Info>? = null

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as ThreadAttachment

            if (title != other.title) return false
            if (infoList != other.infoList) return false

            return true
        }

        override fun hashCode(): Int {
            var result = title?.hashCode() ?: 0
            result = 31 * result + (infoList?.hashCode() ?: 0)
            return result
        }

        @PaperParcel
        @JsonIgnoreProperties(ignoreUnknown = true)
        class Info : PaperParcelable {

            @JsonIgnore
            @get:JsonGetter
            var label: String? = null

            @JsonIgnore
            @get:JsonGetter
            var value: String? = null

            constructor()

            @JsonCreator
            constructor(@JsonProperty("title") label: String?,
                        @JsonProperty("value") value: String?,
                        @JsonProperty("unit") unit: String?) {
                this.label = label
                this.value = StringUtil.unescapeNonBreakingSpace(value) + StringUtils.defaultString(unit)
            }

            override fun equals(other: Any?): Boolean {
                if (this === other) return true
                if (javaClass != other?.javaClass) return false

                other as Info

                if (label != other.label) return false
                if (value != other.value) return false

                return true
            }

            override fun hashCode(): Int {
                var result = label?.hashCode() ?: 0
                result = 31 * result + (value?.hashCode() ?: 0)
                return result
            }

            companion object {
                @JvmField
                val CREATOR = PaperParcelPosts_ThreadAttachment_Info.CREATOR
            }
        }
    }

    companion object {

        /**
         * @see .filterPost
         */
        @WorkerThread
        fun filterPostList(oPosts: List<Post>?): List<Post> {
            val posts = ArrayList<Post>()
            if (oPosts != null) {
                val blackWords = BlackWordDbWrapper.instance.getAllNotNormalBlackWord()
                for (post in oPosts) {
                    val fPost = filterPost(post, false, blackWords)
                    if (fPost != null) {
                        posts.add(fPost)
                    }
                }
            }
            return posts
        }

        /**
         * 对数据源进行处理
         *
         *  * 标记黑名单用户
         *  * 回复引用新版替换回老版样式
         *  * 过滤屏蔽词（已屏蔽的对象不会在修改屏蔽词后自动更新，而必须是重新的原始对象）
         *
         * 如果修改了过滤状态，则会返回不同的对象
         */
        @WorkerThread
        fun filterPost(post: Post, clone: Boolean = false, blackWords: List<BlackWord>? = null): Post? {
            var nPost: Post = post
            val blackListWrapper = BlackListDbWrapper.getInstance()
            val blackList = blackListWrapper.getMergedBlackList(post.authorId?.toIntOrNull()
                    ?: -1, post.authorName)
            if (blackList == null || blackList.post == BlackList.NORMAL) {
                if (post.hide == Post.Hide_User) {
                    if (clone) {
                        nPost = post.clone()
                    }
                    nPost.hide = Post.Hide_Normal
                }
            } else if (blackList.post == BlackList.DEL_POST) {
                return null
            } else if (blackList.post == BlackList.HIDE_POST) {
                if (post.hide != Post.Hide_User) {
                    if (clone) {
                        nPost = post.clone()
                    }
                    nPost.hide = Post.Hide_User
                }
                nPost.remark = blackList.remark
            }

            val reply = nPost.reply
            if (reply != null && nPost.hide == Post.Hide_Normal) {
                val mBlackWords = blackWords
                        ?: BlackWordDbWrapper.instance.getAllNotNormalBlackWord()
                mBlackWords.forEach {
                    val word = it.word
                    if (!word.isNullOrEmpty() && it.stat != BlackWord.NORMAL) {
                        if (reply.contains(word, false)) {
                            if (it.stat == BlackWord.DEL) {
                                return null
                            } else if (it.stat == BlackWord.HIDE) {
                                //Only clone if not cloned before
                                if (clone && nPost === post) {
                                    nPost = post.clone()
                                }
                                nPost.hide = Post.Hide_Word
                                return@forEach
                            }
                        }
                    }
                }
            }

            return nPost
        }
    }


}
